src/lib/OS/SyscallHandler.cc

```diff
@@ -1,15 +1,756 @@
+      int64_t whence = currentInfo_.registerArguments[2].get<uint64_t>();
+      stateChange = {ChangeType::REPLACEMENT,
+                     {currentInfo_.ret},
+                     {lseek(fd, offset, whence)}};
+      break;
+    }
+    case 63: {  // read
+      int64_t fd = currentInfo_.registerArguments[0].get<int64_t>();
+      uint64_t bufPtr = currentInfo_.registerArguments[1].get<uint64_t>();
+      uint64_t count = currentInfo_.registerArguments[2].get<uint64_t>();
+
+      return readBufferThen(bufPtr, count, [=]() {
+        int64_t totalRead = read(fd, dataBuffer_.data(), count);
+        ProcessStateChange stateChange = {
+            ChangeType::REPLACEMENT, {currentInfo_.ret}, {totalRead}};
+        // Check for failure
+        if (totalRead < 0) {
+          return concludeSyscall(stateChange);
+        }
+
+        int64_t bytesRemaining = totalRead;
+        // Get pointer and size of the buffer
+        uint64_t iDst = bufPtr;
+        uint64_t iLength = bytesRemaining;
+        if (iLength > bytesRemaining) {
+          iLength = bytesRemaining;
+        }
+        bytesRemaining -= iLength;
+
+        // Write data for this buffer in 128-byte chunks
+        auto iSrc = dataBuffer_.data();
+        while (iLength > 0) {
+          uint8_t len = iLength > 128 ? 128 : static_cast<uint8_t>(iLength);
+          stateChange.memoryAddresses.push_back({iDst, len});
+          stateChange.memoryAddressValues.push_back({iSrc, len});
+          iDst += len;
+          iSrc += len;
+          iLength -= len;
+        }
+        concludeSyscall(stateChange);
+      });
+      break;
+    }
+    case 64: {  // write
+      int64_t fd = currentInfo_.registerArguments[0].get<int64_t>();
+      uint64_t bufPtr = currentInfo_.registerArguments[1].get<uint64_t>();
+      uint64_t count = currentInfo_.registerArguments[2].get<uint64_t>();
+
+      return readBufferThen(bufPtr, count, [=]() {
+        int64_t retval = write(fd, dataBuffer_.data(), count);
+        ProcessStateChange stateChange = {
+            ChangeType::REPLACEMENT, {currentInfo_.ret}, {retval}};
+        concludeSyscall(stateChange);
+      });
+      break;
+    }
+    case 65: {  // readv
+      int64_t fd = currentInfo_.registerArguments[0].get<int64_t>();
+      uint64_t iov = currentInfo_.registerArguments[1].get<uint64_t>();
+      int64_t iovcnt = currentInfo_.registerArguments[2].get<int64_t>();
+
+      // The pointer `iov` points to an array of structures that each contain
+      // a pointer to where the data should be written and the number of
+      // bytes to write.
+      //
+      // We're going to queue up two handlers:
+      // - First, read the iovec structures that describe each buffer.
+      // - Second, invoke the kernel to perform the read operation, and
+      //   generate memory write requests for each buffer.
+
+      // Create the second handler in the chain, which invokes the kernel and
```

Comment:
Which kernel are we talking about? host kernel? or SimOS? If it is the simOS should it be named kernel

Reply:
Neither, a kernel as in a code kernel. Whilst it had caused confusion here, I'd argue that its a suitable usage of the term "kernel"

Reply:
What does a code kernel mean?

Reply:
A code kernel refers to a codes main functionality. I think the comment is fine as is